oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticSpellcheckProcess.java

```diff
@@ -0,0 +1,164 @@
+        int i = 0;
+        for (String field : getSpellCheckFields(defn)) {
+            PhraseSuggestionBuilder.CandidateGenerator candidateGeneratorBuilder = new DirectCandidateGeneratorBuilder(field + ".trigram")
+                    .suggestMode("missing");
+            SuggestionBuilder phraseSuggestionBuilder = SuggestBuilders.phraseSuggestion(field + ".trigram")
+                    .size(10)
+                    .addCandidateGenerator(candidateGeneratorBuilder)
+                    .text(spellcheckQueryString)
+                    .collateQuery(getCollateQuery(field, rowIteratorState.plan).toString()).collatePrune(true);
+            suggestBuilder.addSuggestion("cqsuggestion" + i, phraseSuggestionBuilder);
+        }
+        return suggestBuilder;
+    }
+
+    @Override
+    public SearchHit process() throws IOException {
+
+        SearchHit lastDocToRecord = null;
+        rowIteratorState.noDocs = true;
+        ElasticSearcher searcher = new ElasticSearcher(rowIteratorState.indexNode);//getCurrentElasticSearcher(indexNode);
+        SuggestBuilder suggestBuilder = getSuggestBuilder();
+        ElasticIndexDefinition indexDefinition = (ElasticIndexDefinition) rowIteratorState.planResult.indexDefinition;
+
+        ElasticSearcherModel elasticSearcherModel = new ElasticSearcherModel.ElasticSearcherModelBuilder()
+                .withSpellCheck(suggestBuilder).build();
+        SearchResponse docs = searcher.search(elasticSearcherModel);
+        Suggest suggest = docs.getSuggest();
+        // Priority queue to get sorted results with decreasing score
+        Queue<Suggest.Suggestion.Entry.Option> pqueue = new PriorityQueue<>((o1, o2) -> Float.compare(o2.getScore(), o1.getScore()));
+
+        Iterator<Suggest.Suggestion<? extends Suggest.Suggestion.Entry<? extends Suggest.Suggestion.Entry.Option>>> suggestionIterator = suggest.iterator();
+        while (suggestionIterator.hasNext()) {
+            Suggest.Suggestion<? extends Suggest.Suggestion.Entry<? extends Suggest.Suggestion.Entry.Option>> spellCheckResults = suggestionIterator.next();
+            for (Suggest.Suggestion.Entry spell : spellCheckResults) {
+                List<Suggest.Suggestion.Entry.Option> options = spell.getOptions();
+                for (Suggest.Suggestion.Entry.Option option : options) {
+                    if (option.collateMatch()) {
+                        pqueue.add(option);
+                    }
+                }
+            }
+        }
+
+        List<ElasticSearcherModel> elasticSearcherModels = new LinkedList<>();
+        for (Suggest.Suggestion.Entry.Option suggestionoption : pqueue) {
+            String suggestion = suggestionoption.getText().string();
+            List<QueryBuilder> qbList = new LinkedList<>();
+            QueryBuilder queryBuilder = new MultiMatchQueryBuilder(suggestion, getSpellCheckFields(indexDefinition)
+                    .toArray(new String[getSpellCheckFields(indexDefinition).size()]))
```

Comment:
```suggestion
                    .toArray(new String[0]))
```
For an in-depth analysis of why we should do this look at https://shipilev.net/blog/2016/arrays-wisdom-ancients/

Reply:
Doesn't blog suggest to use definite size instead of zero sized array and then resizing?

Reply:
For small size collections the `T[0]` seems to be faster than `T[size]`. From the blog

```
Bottom line: toArray(new T[0]) seems faster, safer, and contractually cleaner, and therefore should be the default choice now. 
```

In our case we are also calling `getSpellCheckFields(indexDefinition)` so `T[0]` will for sure be faster and cleaner.

Reply:
done